File: src/main/java/link/infra/packwiz/installer/ui/OptionsSelectWindow.java
package link.infra.packwiz.installer.ui;

import javax.swing.*;
import javax.swing.border.EmptyBorder;
import javax.swing.event.ListSelectionEvent;
import javax.swing.event.ListSelectionListener;
import javax.swing.event.TableModelListener;
import javax.swing.table.TableModel;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.WindowAdapter;
import java.awt.event.WindowEvent;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;

public class OptionsSelectWindow extends JDialog implements ActionListener {

	private static final long serialVersionUID = 1L;
	private final JTextArea lblOptionDescription;
	private final OptionTableModel tableModel;
	private final CompletableFuture<Boolean> future;

	/**
	 * Create the dialog.
	 */
	OptionsSelectWindow(List<IOptionDetails> optList, CompletableFuture<Boolean> future) {
		tableModel = new OptionTableModel(optList);
		this.future = future;

		setModal(true);
		setTitle("Select optional mods...");
		setBounds(100, 100, 450, 300);
		getContentPane().setLayout(new BorderLayout());
		JPanel contentPanel = new JPanel();
		contentPanel.setBorder(new EmptyBorder(5, 5, 5, 5));
		getContentPane().add(contentPanel, BorderLayout.CENTER);
		contentPanel.setLayout(new BorderLayout(0, 0));
		{
			JSplitPane splitPane = new JSplitPane();
			splitPane.setResizeWeight(0.5);
			contentPanel.add(splitPane);
			{
				JTable table = new JTable();
				table.setShowVerticalLines(false);
				table.setShowHorizontalLines(false);
				table.setSelectionMode(ListSelectionModel.SINGLE_SELECTION);
				table.setShowGrid(false);
				table.setModel(tableModel);
				table.getColumnModel().getColumn(0).setResizable(false);
				table.getColumnModel().getColumn(0).setPreferredWidth(15);
				table.getColumnModel().getColumn(0).setMaxWidth(15);
				table.getColumnModel().getColumn(1).setResizable(false);
				table.getSelectionModel().addListSelectionListener(new ListSelectionListener() {
					@Override
					public void valueChanged(ListSelectionEvent e) {
						int i = table.getSelectedRow();
						if (i > -1) {
							lblOptionDescription.setText(tableModel.getDescription(i));
						} else {
							lblOptionDescription.setText("Select an option...");
						}
					}
				});
				table.setTableHeader(null);
				JScrollPane scrollPane = new JScrollPane(table);
				scrollPane.getViewport().setBackground(UIManager.getColor("List.background"));
				scrollPane.setHorizontalScrollBarPolicy(ScrollPaneConstants.HORIZONTAL_SCROLLBAR_NEVER);
				scrollPane.setBorder(new EmptyBorder(0, 0, 0, 0));
				splitPane.setLeftComponent(scrollPane);
			}
			{
				lblOptionDescription = new JTextArea("Select an option...");
				lblOptionDescription.setBackground(UIManager.getColor("List.background"));
				lblOptionDescription.setOpaque(true);
				lblOptionDescription.setWrapStyleWord(true);
				lblOptionDescription.setLineWrap(true);
				lblOptionDescription.setOpaque(true);
				lblOptionDescription.setEditable(false);
				lblOptionDescription.setFocusable(false);
				lblOptionDescription.setFont(UIManager.getFont("Label.font"));
				lblOptionDescription.setBorder(new EmptyBorder(10, 10, 10, 10));
				JScrollPane scrollPane = new JScrollPane(lblOptionDescription);
				scrollPane.setBorder(new EmptyBorder(0, 0, 0, 0));
				splitPane.setRightComponent(scrollPane);
			}
		}
		{
			JPanel buttonPane = new JPanel();
			buttonPane.setLayout(new FlowLayout(FlowLayout.RIGHT));
			getContentPane().add(buttonPane, BorderLayout.SOUTH);
			{
				JButton okButton = new JButton("OK");
				okButton.setActionCommand("OK");
				okButton.addActionListener(this);
				buttonPane.add(okButton);
				getRootPane().setDefaultButton(okButton);
			}
			{
				JButton cancelButton = new JButton("Cancel");
				cancelButton.setActionCommand("Cancel");
				cancelButton.addActionListener(this);
				buttonPane.add(cancelButton);
			}
		}
		addWindowListener(new WindowAdapter() {
			@Override
			public void windowClosing(WindowEvent e) {
				future.complete(true);
			}

			@Override
			public void windowClosed(WindowEvent e) {
				// Just in case closing didn't get triggered - if something else called dispose() the
				// future will have already completed
				future.complete(true);
			}
		});
	}

	private static class OptionTableModel implements TableModel {
		private List<OptionTempHandler> opts = new ArrayList<>();

		OptionTableModel(List<IOptionDetails> givenOpts) {
			for (IOptionDetails opt : givenOpts) {
				opts.add(new OptionTempHandler(opt));
			}
		}

		@Override
		public int getRowCount() {
			return opts.size();
		}

		@Override
		public int getColumnCount() {
			return 2;
		}

		private final String[] columnNames = {"Enabled", "Mod name"};
		private final Class<?>[] columnTypes = {Boolean.class, String.class};
		private final boolean[] columnEditables = {true, false};

		@Override
		public String getColumnName(int columnIndex) {
			return columnNames[columnIndex];
		}

		@Override
		public Class<?> getColumnClass(int columnIndex) {
			return columnTypes[columnIndex];
		}

		@Override
		public boolean isCellEditable(int rowIndex, int columnIndex) {
			return columnEditables[columnIndex];
		}

		@Override
		public Object getValueAt(int rowIndex, int columnIndex) {
			OptionTempHandler opt = opts.get(rowIndex);
			return columnIndex == 0 ? opt.getOptionValue() : opt.getName();
		}

		@Override
		public void setValueAt(Object aValue, int rowIndex, int columnIndex) {
			if (columnIndex == 0) {
				OptionTempHandler opt = opts.get(rowIndex);
				opt.setOptionValue((boolean) aValue);
			}
		}

		// Noop, the table model doesn't change!
		@Override
		public void addTableModelListener(TableModelListener l) {}

		@Override
		public void removeTableModelListener(TableModelListener l) {}

		String getDescription(int rowIndex) {
			return opts.get(rowIndex).getOptionDescription();
		}

		void finalise() {
			for (OptionTempHandler opt : opts) {
				opt.finalise();
			}
		}

	}

	@Override
	public void actionPerformed(ActionEvent e) {
		if (e.getActionCommand().equals("OK")) {
			tableModel.finalise();
			future.complete(false);
			dispose();
		} else if (e.getActionCommand().equals("Cancel")) {
			future.complete(true);
			dispose();
		}
	}

}

File: src/main/kotlin/link/infra/packwiz/installer/ui/gui/OptionsSelectWindow.kt
package link.infra.packwiz.installer.ui.gui

import link.infra.packwiz.installer.ui.data.IOptionDetails
import java.awt.BorderLayout
import java.awt.FlowLayout
import java.awt.event.ActionEvent
import java.awt.event.ActionListener
import java.awt.event.WindowAdapter
import java.awt.event.WindowEvent
import java.util.*
import java.util.concurrent.CompletableFuture
import javax.swing.*
import javax.swing.border.EmptyBorder
import javax.swing.event.TableModelListener
import javax.swing.table.TableModel

class OptionsSelectWindow internal constructor(optList: List<IOptionDetails>, future: CompletableFuture<Boolean>, parentWindow: JFrame?) : JDialog(parentWindow, "Select optional mods...", true), ActionListener {
	private val lblOptionDescription: JTextArea
	private val tableModel: OptionTableModel
	private val future: CompletableFuture<Boolean>

	private class OptionTableModel(givenOpts: List<IOptionDetails>) : TableModel {
		private val opts: List<OptionTempHandler>

		init {
			val mutOpts = ArrayList<OptionTempHandler>()
			for (opt in givenOpts) {
				mutOpts.add(OptionTempHandler(opt))
			}
			opts = mutOpts
		}

		override fun getRowCount() = opts.size
		override fun getColumnCount() = 2

		private val columnNames = arrayOf("Enabled", "Mod name")
		private val columnTypes = arrayOf(Boolean::class.javaObjectType, String::class.java)
		private val columnEditables = booleanArrayOf(true, false)

		override fun getColumnName(columnIndex: Int) = columnNames[columnIndex]
		override fun getColumnClass(columnIndex: Int) = columnTypes[columnIndex]
		override fun isCellEditable(rowIndex: Int, columnIndex: Int) = columnEditables[columnIndex]

		override fun getValueAt(rowIndex: Int, columnIndex: Int): Any {
			val opt = opts[rowIndex]
			return if (columnIndex == 0) opt.optionValue else opt.name
		}

		override fun setValueAt(aValue: Any, rowIndex: Int, columnIndex: Int) {
			if (columnIndex == 0) {
				val opt = opts[rowIndex]
				opt.optionValue = aValue as Boolean
			}
		}

		// Noop, the table model doesn't change!
		override fun addTableModelListener(l: TableModelListener) {}
		override fun removeTableModelListener(l: TableModelListener) {}

		fun getDescription(rowIndex: Int) = opts[rowIndex].optionDescription

		fun finalise() {
			for (opt in opts) {
				opt.finalise()
			}
		}
	}

	override fun actionPerformed(e: ActionEvent) {
		if (e.actionCommand == "OK") {
			tableModel.finalise()
			future.complete(false)
			dispose()
		} else if (e.actionCommand == "Cancel") {
			future.complete(true)
			dispose()
		}
	}

	/**
	 * Create the dialog.
	 */
	init {
		tableModel = OptionTableModel(optList)
		this.future = future

		setBounds(100, 100, 450, 300)
		setLocationRelativeTo(parentWindow)

		contentPane.apply {
			layout = BorderLayout()
			add(JPanel().apply {
				border = EmptyBorder(5, 5, 5, 5)
				layout = BorderLayout(0, 0)

				add(JSplitPane().apply {
					resizeWeight = 0.5

					lblOptionDescription = JTextArea("Select an option...").apply {
						background = UIManager.getColor("List.background")
						isOpaque = true
						wrapStyleWord = true
						lineWrap = true
						isEditable = false
						isFocusable = false
						font = UIManager.getFont("Label.font")
						border = EmptyBorder(10, 10, 10, 10)
					}

					leftComponent = JScrollPane(JTable().apply {
						showVerticalLines = false
						showHorizontalLines = false
						setSelectionMode(ListSelectionModel.SINGLE_SELECTION)
						setShowGrid(false)
						model = tableModel
						columnModel.getColumn(0).resizable = false
						columnModel.getColumn(0).preferredWidth = 15
						columnModel.getColumn(0).maxWidth = 15
						columnModel.getColumn(1).resizable = false
						selectionModel.addListSelectionListener {
							val i = selectedRow
							if (i > -1) {
								lblOptionDescription.text = tableModel.getDescription(i)
							} else {
								lblOptionDescription.text = "Select an option..."
							}
						}
						tableHeader = null
					}).apply {
						viewport.background = UIManager.getColor("List.background")
						horizontalScrollBarPolicy = ScrollPaneConstants.HORIZONTAL_SCROLLBAR_NEVER
					}

					rightComponent = JScrollPane(lblOptionDescription)
				})

				add(JPanel().apply {
					layout = FlowLayout(FlowLayout.RIGHT)

					add(JButton("OK").apply {
						actionCommand = "OK"
						addActionListener(this@OptionsSelectWindow)

						this@OptionsSelectWindow.rootPane.defaultButton = this
					})

					add(JButton("Cancel").apply {
						actionCommand = "Cancel"
						addActionListener(this@OptionsSelectWindow)
					})
				}, BorderLayout.SOUTH)
			}, BorderLayout.CENTER)
		}

		addWindowListener(object : WindowAdapter() {
			override fun windowClosing(e: WindowEvent) {
				future.complete(true)
			}

			override fun windowClosed(e: WindowEvent) {
				// Just in case closing didn't get triggered - if something else called dispose() the
				// future will have already completed
				future.complete(true)
			}
		})
	}
}